lib/mongoid/clients/options.rb

```diff
@@ -93,6 +107,97 @@ def persistence_options(klass = self)
         def set_persistence_options(klass, options)
           Thread.current["[mongoid][#{klass}]:persistence-options"] = options
         end
+
+        # Unset the persistence options on the current thread.
+        #
+        # @api private
+        #
+        # @example Unset the persistence options.
+        #   Threaded.unset_persistence_options(Band)
+        #
+        # @param [ Class ] klass The model class.
+        #
+        # @return [ nil ] nil.
+        #
+        # @since 5.1.0
+        def unset_persistence_options(klass)
+          Thread.current["[mongoid][#{klass}]:persistence-options"] = nil
+        end
+
+        # Set the persistence options and client with those options on the current thread.
+        # Note that a client will only be set if its cluster differs from the cluster of the
+        # original client.
+        #
+        # @api private
+        #
+        # @example Set the persistence options and client with those options on the current thread.
+        #   Threaded.set_options(Band, { write: { w: 3 }})
+        #
+        # @param [ Class ] klass The model class.
+        # @param [ Mongo::Client ] client The client with options.
+        #
+        # @return [ Mongo::Client, nil ] The client or nil if the cluster does not change.
+        #
+        # @since 5.1.0
+        def set_options(klass, options)
+          original_cluster = mongo_client.cluster
+          set_persistence_options(klass, options)
+          m = mongo_client
+          set_client_with_options(klass, m) unless m.cluster == original_cluster
+        end
+
+        # Set the client with special options on the current thread.
+        #
+        # @api private
+        #
+        # @example Set the client with options.
+        #   Threaded.set_client_with_options(Band, client)
+        #
+        # @param [ Class ] klass The model class.
+        # @param [ Mongo::Client ] client The client with options.
+        #
+        # @return [ Mongo::Client ] The client.
+        #
+        # @since 5.1.0
+        def set_client_with_options(klass, client)
+          Thread.current["[mongoid][#{klass}]:mongo-client"] = client
+        end
+
+        # Unset the client with special options on the current thread.
+        #
+        # @api private
+        #
+        # @example Unset the client with options.
+        #   Threaded.unset_client_with_options(Band)
+        #
+        # @param [ Class ] klass The model class.
+        #
+        # @return [ nil ] nil.
+        #
+        # @since 5.1.0
+        def unset_client_with_options(klass)
+          if Thread.current["[mongoid][#{klass}]:mongo-client"]
```

Comment:
sure
Otherwise, do you have another idea of how to address this? I think the solution could be cleaner so will continue thinking about it / iterating on it.